AsyncDisplayKit/Details/ASRangeControllerBeta.mm

```diff
@@ -106,21 +141,25 @@ - (void)_updateVisibleNodeIndexPaths
   NSMutableOrderedSet<NSIndexPath *> *allIndexPaths = [[NSMutableOrderedSet alloc] initWithSet:visibleIndexPaths];
   
   ASInterfaceState selfInterfaceState = [_dataSource interfaceStateForRangeController:self];
-  BOOL selfIsVisible = (ASInterfaceStateIncludesVisible(selfInterfaceState));
-  BOOL selfIsScrolling = (_scrollDirection != ASScrollDirectionNone);
-  // If we are already visible and scrolling, get busy!  Better get started on preloading before the user scrolls more...
-  // If we used full range, don't switch to minimum range now. That will destroy all the hard work done before.
-  BOOL shouldUseFullRange = ((selfIsVisible && selfIsScrolling) || _didUseFullRange);
+  ASLayoutRangeMode rangeMode = [ASRangeControllerBeta rangeModeForInterfaceState:selfInterfaceState
+                                                                  scrollDirection:_scrollDirection
+                                                                 currentRangeMode:_currentRangeMode];
 
-  fetchDataIndexPaths = [_layoutController indexPathsForScrolling:_scrollDirection rangeType:ASLayoutRangeTypeFetchData shouldUseFullRange:shouldUseFullRange];
+  fetchDataIndexPaths = [_layoutController indexPathsForScrolling:_scrollDirection
+                                                        rangeMode:rangeMode
+                                                        rangeType:ASLayoutRangeTypeFetchData];
 
-  ASRangeTuningParameters parametersDisplay = [_layoutController tuningParametersForRangeType:ASLayoutRangeTypeDisplay isFullRange:shouldUseFullRange];
-  ASRangeTuningParameters parametersFetchData = [_layoutController tuningParametersForRangeType:ASLayoutRangeTypeFetchData isFullRange:shouldUseFullRange];
+  ASRangeTuningParameters parametersDisplay = [_layoutController tuningParametersForRangeMode:rangeMode
+                                                                                    rangeType:ASLayoutRangeTypeDisplay];
+  ASRangeTuningParameters parametersFetchData = [_layoutController tuningParametersForRangeMode:rangeMode
+                                                                                      rangeType:ASLayoutRangeTypeFetchData];
   if (parametersDisplay.leadingBufferScreenfuls == parametersFetchData.leadingBufferScreenfuls &&
```

Comment:
I wonder if there is an easy way to generalize this? We should check if either range is equal to the visible range, As a developer could very reasonably choose to set their minimum display and a fetch data ranges to zero especially if they're Collection view layout is expensive and they need to minimize calls to get the index paths that lie within a given range